### PR TITLE
chore: enforce test coverage thresholds

### PR DESCRIPTION
**Pull request by LayZeeDK**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[x] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
No automatically enforced limits for test coverage.

Issue Number: N/A

## What is the new behavior?
- **Minimum test coverage of 85%** in demo application and Lumberjack library, verified by the CI workflow.
- Test coverage reports archived as build artifacts.
- Test coverage reported in PR description and compared to `main` branch after each commit.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Other information
- Covered files under test coverage threshold.
- Simplified CI workflow by removing matrices with only one leg.
- Corrected 2 CI workflow jobs.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -198,3 +186,122 @@ jobs:
...
+      # consolidate after PR#66 has been merged
+      - if: github.ref != 'refs/pull/66/merge'
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event.pull_request.head.ref }}
+          fetch-depth: 0
+      - if: github.ref != 'refs/pull/66/merge'
+        run: git fetch --no-tags --prune --depth=5 origin main
+      - if: github.ref == 'refs/pull/66/merge'
+        uses: actions/checkout@v2
```
Comment (NachoVazquez) [inline .github/workflows/ci.yml:205]: Can you explain what is this doing and why?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -198,3 +186,122 @@ jobs:
...
+      - name: Cache Yarn cache directory
+        uses: actions/cache@v2
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-node-${{ env.NODE_LATEST }}-angular-${{ env.ANGULAR_LATEST }}-yarn-${{ hashFiles('yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-node-${{ env.NODE_LATEST }}-angular-${{ env.ANGULAR_LATEST }}-yarn-
+            ${{ runner.os }}-node-${{ env.NODE_LATEST }}-angular-
+      - run: yarn install
+      - run: npx ngcc || true
```
Comment (NachoVazquez) [inline .github/workflows/ci.yml:224]: Is this necessary for all Angular versions or just for some?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -198,3 +186,122 @@ jobs:
...
+
+      - run: yarn test:lib:coverage #--no-fail-on-failing-test-suite
+      - name: Rename @ngworker/lumberjack base test coverage report
+        run: mv coverage/libs/ngworker/lumberjack/lcov.info coverage/libs/ngworker/lumberjack/lcov-base.info
```
Comment (NachoVazquez) [inline .github/workflows/ci.yml:228]: Why is this needed?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -198,3 +186,122 @@ jobs:
...
+          if-no-files-found: error
+      - uses: actions/download-artifact@v2
+        with:
+          name: lumberjack-app-base-coverage
+          path: coverage/apps/lumberjack-app
+
+      - run: mv coverage/libs/ngworker/* coverage
+      - run: mv coverage/apps/* coverage
+      - name: Report test coverage in PR description
+        uses: ScaCap/code-coverage-assistant@v0.1
```
Comment (NachoVazquez) [inline .github/workflows/ci.yml:304]: This is in v1, why use an old version?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -198,3 +186,122 @@ jobs:
           sudo apt-get --only-upgrade install google-chrome-stable
           google-chrome --version
       - run: yarn e2e
+
+  coverage-base:
+    if: github.event_name == 'pull_request'
```
Comment (NachoVazquez) [inline .github/workflows/ci.yml:191]: What's the purpose of this job? Why it is different from the **coverage** job.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (LayZeeDK) [reply .github/workflows/ci.yml:205]: It's checking out the base branch to calculate current test coverage.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (LayZeeDK) [reply .github/workflows/ci.yml:224]: Because of fucking ngcc 😒 CI messed up for Angular 9.0 or 9.x because it was such a terrible release where ngcc behavior changed a hundred times.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (LayZeeDK) [reply .github/workflows/ci.yml:228]: Because the coverage action expects a report of this name and I'm trying to avoid having 2 karma.conf.js files per project.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (LayZeeDK) [reply .github/workflows/ci.yml:191]: We're checking out the base branch so that the coverage action can compare the test coverage impact of merging this branch.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (LayZeeDK) [reply .github/workflows/ci.yml:205]: We should remove the hard-coded main branch name somehow. It might not even be necessary, but that's hard to tell until the next PR.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (NachoVazquez) [reply .github/workflows/ci.yml:224]: Yeah, but you are only running this job against the latest Angular version, that's why I asked.

Comment (LayZeeDK) [reply .github/workflows/ci.yml:224]: That's a good point. It think it's still useful to run it beforehand so that is clear which step is failing.

Comment (LayZeeDK) [reply .github/workflows/ci.yml:224]: The || true part was because of View Engine though.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (NachoVazquez) [reply .github/workflows/ci.yml:191]: That was my guest, but where I can see that calculation in the report? I can only see the resulting coverage report after merging? 

I don't see anything that says "...you improve x %..." or "coverage before merge is ..."

Comment (LayZeeDK) [reply .github/workflows/ci.yml:191]: It says 0.00% change. Because of the temporary branch conditional, it reports coverage on the same branch instead of the `main` branch since test coverage reports are currently not enabled in `main`. I'll try branching out from this one and creating a PR against this one to try the intended conditional branch and to get rid of the hardcoded `main` in one of the steps. It won't show any difference in tests unless I cover some of the remaining percentages in that new branch.

**Review by NachoVazquez — APPROVED** [review]

**Comment by LayZeeDK** [comment]
I'm adding SonarCloud instead. It will save us some CI execution time, it will add more analyses, it will keep track of code quality metrics over time and it will show the impact of merging a PR on code quality. It will also show up as a GitHub *check*.

**Review by santoshyadavdev — APPROVED** [review]
LGTM

**Review by SerkanSipahi — APPROVED** [review]
LGTM

**Comment by LayZeeDK** [comment]
Thanks @santoshyadavdev @SerkanSipahi, but as mentioned in a comment, I'm abandoning this PR (which is a draft by the way 🙂) in favor of https://github.com/ngworker/lumberjack/pull/68.

**Comment by LayZeeDK** [comment]
Closed in favor of #68.
